Change doc for `google_compute_instance` to reflect the correct behaviour of `subnetwork_project` field

karolgorc opened the pull request:
<!-- Put a description of what this PR is for here, along with any references to issues that this resolves or contributes to -->
closes

The `subnetwork_project` field is ignored in the API request when a self_link is given in `subnetwork`, but not in terraform which is correct behavior because `subnetwork_project` is still in the code. Suppressing a diff on this seems unreasonable because providing something like this isn't really a valid config
```
subnetwork_project = "projectA"
subnetwork = "...projectB..."
```

Maybe throw an error when this happens?

<!--
Please self-review your PR against the review checklist before creating it: https://googlecloudplatform.github.io/magic-modules/contribute/review-pr/

Completing the checklist will help speed up the review process, and we appreciate you spending time on them before sending
your code to be reviewed.

If your PR is still work in progress, please create it in draft mode
-->

<!-- AUTOCHANGELOG for Downstream PRs.

Please select one of the following "release-note:" headings:
    - release-note:enhancement
    - release-note:bug
    - release-note:note
    - release-note:new-resource
    - release-note:new-datasource
    - release-note:deprecation
    - release-note:breaking-change
    - release-note:none

Unless you choose release-note:none, please add a release note.

See https://googlecloudplatform.github.io/magic-modules/contribute/release-notes/ for writing good release notes.

You can add more release note blocks if you want more than one CHANGELOG
entry for this PR.
-->
**Release Note Template for Downstream PRs (will be copied)**

```release-note:note
compute: change doc for `google_compute_instance` `subnetwork_project`
```

zli82016 commented:
/gcbrun

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report

Your PR generated some diffs in downstreams - here they are.

`google` provider: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-11537-old..auto-pr-11537) ( 1 file changed, 1 insertion(+), 1 deletion(-))
`google-beta` provider: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-11537-old..auto-pr-11537) ( 1 file changed, 1 insertion(+), 1 deletion(-))

karolgorc commented:
that unit-test fail doesn't seem related

zli82016 commented:
> The `subnetwork_project` field is ignored in the API request when a self_link is given in `subnetwork`, but not in terraform which is correct behavior because `subnetwork_project` is still in the code. Suppressing a diff on this seems unreasonable because providing something like this isn't really a valid config
> 
> ```
> subnetwork_project = "projectA"
> subnetwork = "...projectB..."
> ```
> 
> Maybe throw an error when this happens?

Throwing an error is a good idea to let the user know that the provided subnetwork_project is not correct.

zli82016 reviewed (changes requested):
Can you please add the validation? Thanks.

karolgorc commented:
- Added GetProjectFromRegionalSelfLink to tpgresource and a unit test for it
- Added the new logic to _CustomizeDiff_

This isn't really a simple field validation to use _ValidateFunc_ because it depends on an external value which isn't really supported by the types used in _ValidateFunc_, _StateFunc_ or _DiffSuppressFunc_.

![image](https://github.com/user-attachments/assets/083323bd-42ca-446d-bed8-9f5c8d63f757)

zli82016 commented:
/gcbrun

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report

Your PR generated some diffs in downstreams - here they are.

`google` provider: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-11537-old..auto-pr-11537) ( 4 files changed, 55 insertions(+), 1 deletion(-))
`google-beta` provider: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-11537-old..auto-pr-11537) ( 4 files changed, 55 insertions(+), 1 deletion(-))

modular-magician commented:
#### Tests analytics
Total tests: 3921
Passed tests: 3509
Skipped tests: 411
Affected tests: 1

<details>
<summary>Click here to see the affected service packages</summary>
<blockquote>

All service packages are affected

</blockquote>
</details>
#### Non-exercised tests

Tests were added that are skipped in VCR:
- TestGetProjectFromRegionalSelfLink
#### Action taken
<details>
<summary>Found 1 affected test(s) by replaying old test recordings. Starting RECORDING based on the most recent commit. Click here to see the affected tests
</summary>
<blockquote>
<ul>
<li>TestAccComputeRegionPerInstanceConfig_removeInstanceOnDestroy</li>

</ul>
</blockquote>
</details>

[Get to know how VCR tests work](https://googlecloudplatform.github.io/magic-modules/docs/getting-started/contributing/#general-contributing-steps)

modular-magician commented:
$\textcolor{green}{\textsf{Tests passed during RECORDING mode:}}$
`TestAccComputeRegionPerInstanceConfig_removeInstanceOnDestroy`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-11537/artifacts/a8945e85-9121-4cba-a813-ef53963d6f09/recording/TestAccComputeRegionPerInstanceConfig_removeInstanceOnDestroy.log)]

$\textcolor{green}{\textsf{No issues found for passed tests after REPLAYING rerun.}}$

---




$\textcolor{green}{\textsf{All tests passed!}}$

View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-11537/artifacts/a8945e85-9121-4cba-a813-ef53963d6f09/build-log/recording_test.log) or the [debug log](https://console.cloud.google.com/storage/browser/ci-vcr-logs/beta/refs/heads/auto-pr-11537/artifacts/a8945e85-9121-4cba-a813-ef53963d6f09/recording) for each test

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report

Your PR generated some diffs in downstreams - here they are.

`google` provider: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-11537-old..auto-pr-11537) ( 4 files changed, 55 insertions(+), 1 deletion(-))
`google-beta` provider: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-11537-old..auto-pr-11537) ( 4 files changed, 55 insertions(+), 1 deletion(-))

modular-magician commented:
#### Tests analytics
Total tests: 3923
Passed tests: 3510
Skipped tests: 411
Affected tests: 2

<details>
<summary>Click here to see the affected service packages</summary>
<blockquote>

All service packages are affected

</blockquote>
</details>
#### Non-exercised tests

Tests were added that are skipped in VCR:
- TestGetProjectFromRegionalSelfLink
#### Action taken
<details>
<summary>Found 2 affected test(s) by replaying old test recordings. Starting RECORDING based on the most recent commit. Click here to see the affected tests
</summary>
<blockquote>
<ul>
<li>TestAccComputeInstanceTemplate_withNamePrefix</li>
<li>TestAccComputeTargetHttpsProxy_update</li>

</ul>
</blockquote>
</details>

[Get to know how VCR tests work](https://googlecloudplatform.github.io/magic-modules/docs/getting-started/contributing/#general-contributing-steps)

modular-magician commented:
$\textcolor{green}{\textsf{Tests passed during RECORDING mode:}}$
`TestAccComputeInstanceTemplate_withNamePrefix`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-11537/artifacts/7735ed4c-6498-4d25-9a67-0956deb4f105/recording/TestAccComputeInstanceTemplate_withNamePrefix.log)]
`TestAccComputeTargetHttpsProxy_update`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-11537/artifacts/7735ed4c-6498-4d25-9a67-0956deb4f105/recording/TestAccComputeTargetHttpsProxy_update.log)]
$\textcolor{red}{\textsf{Tests failed when rerunning REPLAYING mode:}}$
`TestAccComputeInstanceTemplate_withNamePrefix`[[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-11537/artifacts/7735ed4c-6498-4d25-9a67-0956deb4f105/build-log/replaying_build_after_recording/TestAccComputeInstanceTemplate_withNamePrefix_replaying_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-11537/artifacts/7735ed4c-6498-4d25-9a67-0956deb4f105/replaying_after_recording/TestAccComputeInstanceTemplate_withNamePrefix.log)]


Tests failed due to non-determinism or randomness when the VCR replayed the response after the HTTP request was made.

Please fix these to complete your PR. If you believe these test failures to be incorrect or unrelated to your change, or if you have any questions, please raise the concern with your reviewer.


---




$\textcolor{green}{\textsf{All tests passed!}}$

View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-11537/artifacts/7735ed4c-6498-4d25-9a67-0956deb4f105/build-log/recording_test.log) or the [debug log](https://console.cloud.google.com/storage/browser/ci-vcr-logs/beta/refs/heads/auto-pr-11537/artifacts/7735ed4c-6498-4d25-9a67-0956deb4f105/recording) for each test

karolgorc commented:
Got the same test fails in 2 other PR's. Don't think they are related.

zli82016 commented:
Thank you for adding the validation. 

Do you mind adding an acceptance test to check if the incorrect configuration for subnetwork_project will returns error in https://github.com/GoogleCloudPlatform/magic-modules/blob/main/mmv1/third_party/terraform/services/compute/resource_compute_instance_test.go.erb?

Example: https://github.com/GoogleCloudPlatform/magic-modules/blob/main/mmv1/third_party/terraform/services/alloydb/resource_alloydb_cluster_restore_test.go#L39-L42

zli82016 commented:
/gcbrun

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report

Your PR generated some diffs in downstreams - here they are.

`google` provider: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-11537-old..auto-pr-11537) ( 5 files changed, 111 insertions(+), 1 deletion(-))
`google-beta` provider: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-11537-old..auto-pr-11537) ( 5 files changed, 111 insertions(+), 1 deletion(-))

modular-magician commented:
#### Tests analytics
Total tests: 3934
Passed tests: 3520
Skipped tests: 412
Affected tests: 2

<details>
<summary>Click here to see the affected service packages</summary>
<blockquote>

All service packages are affected

</blockquote>
</details>
#### Non-exercised tests

Tests were added that are skipped in VCR:
- TestGetProjectFromRegionalSelfLink
#### Action taken
<details>
<summary>Found 2 affected test(s) by replaying old test recordings. Starting RECORDING based on the most recent commit. Click here to see the affected tests
</summary>
<blockquote>
<ul>
<li>TestAccComputeInstanceTemplate_withNamePrefix</li>
<li>TestAccComputeInstance_subnetworkProjectMustMatchError</li>

</ul>
</blockquote>
</details>

[Get to know how VCR tests work](https://googlecloudplatform.github.io/magic-modules/docs/getting-started/contributing/#general-contributing-steps)

modular-magician commented:
$\textcolor{green}{\textsf{Tests passed during RECORDING mode:}}$
`TestAccComputeInstanceTemplate_withNamePrefix`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-11537/artifacts/18204cab-a75d-4437-9c60-97fb6e08fd29/recording/TestAccComputeInstanceTemplate_withNamePrefix.log)]
`TestAccComputeInstance_subnetworkProjectMustMatchError`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-11537/artifacts/18204cab-a75d-4437-9c60-97fb6e08fd29/recording/TestAccComputeInstance_subnetworkProjectMustMatchError.log)]
$\textcolor{red}{\textsf{Tests failed when rerunning REPLAYING mode:}}$
`TestAccComputeInstanceTemplate_withNamePrefix`[[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-11537/artifacts/18204cab-a75d-4437-9c60-97fb6e08fd29/build-log/replaying_build_after_recording/TestAccComputeInstanceTemplate_withNamePrefix_replaying_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-11537/artifacts/18204cab-a75d-4437-9c60-97fb6e08fd29/replaying_after_recording/TestAccComputeInstanceTemplate_withNamePrefix.log)]


Tests failed due to non-determinism or randomness when the VCR replayed the response after the HTTP request was made.

Please fix these to complete your PR. If you believe these test failures to be incorrect or unrelated to your change, or if you have any questions, please raise the concern with your reviewer.


---




$\textcolor{green}{\textsf{All tests passed!}}$

View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-11537/artifacts/18204cab-a75d-4437-9c60-97fb6e08fd29/build-log/recording_test.log) or the [debug log](https://console.cloud.google.com/storage/browser/ci-vcr-logs/beta/refs/heads/auto-pr-11537/artifacts/18204cab-a75d-4437-9c60-97fb6e08fd29/recording) for each test

zli82016 reviewed (approved):
LGTM. Thanks for adding the validation. The failed tests are unrelated.
